tools/wptrunner/wptrunner/testrunner.py

```diff
@@ -844,12 +900,20 @@ def after_test_end(self, test, restart, force_rerun=False, force_stop=False):
             if subsuite != self.state.subsuite:
                 self.logger.info(f"Restarting browser for new subsuite:{subsuite}")
                 restart = True
-            elif test_type != self.state.test_type:
-                self.logger.info(f"Restarting browser for new test type:{test_type}")
-                restart = True
             elif self.restart_on_new_group and test_group is not self.state.test_group:
                 self.logger.info("Restarting browser for new test group")
                 restart = True
+            elif test_type != self.state.test_type:
+                if self.browser.browser.restart_on_test_type_change(test_type, self.state.test_type):
```

Comment:
Since there's also functional requirements that the `browser` class doesn't know about (i.e. whether the two test types share a protocol) we should actually have a method in this class that does a pre-check before deferring to the browser class iff the pre-check passes.

Reply:
jonathanjlee@ made a similar suggestion previously: https://github.com/web-platform-tests/wpt/pull/40556#discussion_r1608698082.

Looks like this is kind of the same as your comment in `tools/wptrunner/wptrunner/executors/base.py`. How about we check protocol type in `switch_executor`, and if protocol type does not match, we would let `switch_executor` to fail?